File: python/src/zm-s3-upload.py
#!/usr/bin/python3
from boto3 import client
from os import listdir
import sys
import traceback
import logging

objects = ['person', 'bicycle', 'car', 'motorbike', 'bus', 'truck', 'dog', 'cat']


def generate_s3_suffix(description):
    suffix = ''
    if len(description) > 0:
        if description.find('Motion') != -1:
            suffix += '_Motion'
        for object in objects:
            if description.find(object) != -1:
                suffix += f"-{object}"
    return suffix


def generate_s3_path(monitor, date_time):
    split = date_time.split()
    date = split[0]
    time = split[1]
    hours, minutes, seconds = time.split(':')
    return f"{date}/{hours}:{minutes}_{monitor}"


def generate_s3_key(monitor, date_time, description):
    prefix = generate_s3_path(monitor, date_time)
    suffix = generate_s3_suffix(description)
    return f"{prefix}{suffix}.avi"


# unused for now - could possibly use ffmpeg in a lambda instead of forcing ZM zerver to do expensive jpg/avi conversion
def create_video(path, frames, duration):
    frame_rate = round((float(frames) / float(duration)), 2)
    ffmpeg_string = f"ffmpeg -framerate {frame_rate} -pattern_type glob -i '{path}/*capture.jpg' {path}/upload.avi"
    system(ffmpeg_string)


def upload_video(path, monitor, date_time, description):
    s3_client = client('s3')
    files = [f for f in listdir(path) if f.endswith('.avi')]
    if len(files) > 0:
        file_name = f"{path}/{files[0]}"
        bucket = 'zoneminder.mattgilbride.com'
        object_name = generate_s3_key(monitor, date_time, description)
        s3_client.upload_file(file_name, bucket, object_name)
    else:
        logging.error(f"did not find avi file in directory: ${path}")


def upload_raw_zm_files(path, monitor, date_time, description):
    s3_client = client('s3')
    bucket = 'zoneminder.mattgilbride.com'
    s3_path = f"{generate_s3_path(monitor, date_time)}{generate_s3_suffix(description)}"
    for f in listdir(path):
        file_name = f"{path}/{f}"
        object_name = f"{s3_path}/{f}"
        s3_client.upload_file(file_name, bucket, object_name)
    s3_client.put_object(Body=description, Bucket=bucket, Key=f"{s3_path}/-description.txt")


def main(date_time, monitor, description, path):
    logging.basicConfig(filename='/var/log/zm/s3-upload.log', level=logging.INFO)
    logging.info(f"Starting S3 Upload: {path}")
    try:
        # create_video(path, frames, duration)
        upload_raw_zm_files(path, monitor, date_time, description)
        logging.info(f"Finished S3 Upload: {path}")
    except Exception as e:
        track = traceback.format_exc()
        logging.error(track)


if __name__ == '__main__':
    main(sys.argv[1], sys.argv[2], sys.argv[3], sys.argv[4])

File: zm-s3-upload.ts
#!/usr/bin/node
import { S3 } from "aws-sdk";
import fs from 'fs';
import { ManagedUpload } from "aws-sdk/lib/s3/managed_upload";
import SendData = ManagedUpload.SendData;

const generateS3Path = (monitor: string, dateTime: string): string => {
  const [date, time] = dateTime.split(' ')
  const [hours, minutes] = time.split(':')
  return `${date}/${hours}:${minutes}_${monitor}`
}

const objects = ['person', 'bicycle', 'car', 'motorbike', 'bus', 'truck', 'dog', 'cat']

const generateS3Suffix = (description: string): string => {
  let suffix = ''
  if (description && description.length) {
    if (description.search('Motion') > -1) {
      suffix += '_Motion'
    }
    objects.forEach(object => {
      if (description.search(object) > -1) {
        suffix += `-${object}`
      }
    })
  }
  return suffix
}

export interface Manifest {
  expected: number;
  received: number;
  converted: boolean;
  description: string;
  path: string;
  frames: string[]
}

const errorHandler = (err: Error) => {
  console.error(`ERROR: ${err.message}`)
}

const successHandler = (data: SendData) => {
  console.log(`COMPLETED: ${data.Key}`)
}

const uploadRawZmFiles = (dateTime: string, monitor: string, description: string, path: string) => {
  const s3 = new S3()
  const files = fs.readdirSync(path)
  const s3Path = `${generateS3Path(monitor, dateTime)}${generateS3Suffix(description)}/`

  let manifest: Manifest = {
    expected: files.length + 1,
    received: 0,
    converted: false,
    description,
    path: s3Path,
    frames: []
  };

  files.forEach(file => {

    const fileBlob = fs.readFileSync(`${path}/${file}`);
    let objectName = `${s3Path}${file}`

    manifest.received += 1;
    if (file.endsWith('capture.jpg')) {
      objectName = `${s3Path}frames/${file}`
      manifest.frames.push(objectName);
    } else if (file.endsWith('analyze.jpg')) {
      objectName = `${s3Path}motion/${file}`
    }

    s3.upload({
      Bucket: 'zoneminder.mattgilbride.com',
      Body: fileBlob,
      Key: objectName
    })
      .promise()
      .then(successHandler)
      .catch(errorHandler)
  })

  manifest.received += 1;

  s3.upload({
    Bucket: 'zoneminder.mattgilbride.com',
    Body: JSON.stringify(manifest, null, 2),
    Key: `${s3Path}-manifest.json`
  })
    .promise()
    .then(successHandler)
    .catch(errorHandler)
}

const main = () => {
  const args = process.argv.slice(2) as [string, string, string, string]
  uploadRawZmFiles(...args)
}

main()
